README.md
# [medium-editor-textcolor](https://github.com/sonce/medium-editor-textcolor)
[![](https://img.shields.io/badge/Powered%20by-medium-editor-textcolor%20base-brightgreen.svg)](https://github.com/sonce/medium-editor-textcolor)
[![license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/sonce/medium-editor-textcolor/blob/master/LICENSE)
[![Build Status](https://travis-ci.org/sonce/medium-editor-textcolor.svg?branch=master)](https://travis-ci.org/sonce/medium-editor-textcolor)
[![Coveralls](https://img.shields.io/coveralls/sonce/medium-editor-textcolor.svg)](https://coveralls.io/github/sonce/medium-editor-textcolor)
[![npm](https://img.shields.io/badge/npm-0.1.0-orange.svg)](https://www.npmjs.com/package/medium-editor-textcolor)
[![NPM downloads](http://img.shields.io/npm/dm/medium-editor-textcolor.svg?style=flat-square)](http://www.npmtrends.com/medium-editor-textcolor)
[![Percentage of issues still open](http://isitmaintained.com/badge/open/sonce/medium-editor-textcolor.svg)](http://isitmaintained.com/project/sonce/medium-editor-textcolor "Percentage of issues still open")

MediumEditor Text Color is an extension to add a Color Picker button/behavior to MediumEditor.

![Medium Editor TextColor](https://raw.githubusercontent.com/sonce/medium-editor-textcolor/master/demo/images/ui.jpg "TextColor")

Demo: http://sonce.github.io/medium-editor-textcolor/

## Usage

You can install manually or either by using npm or bower:

```
npm install medium-editor-textcolor
```

or

```
bower install medium-editor-textcolor
```

On your app, link the style and the script and initialize MediumEditor with the textcolor extension:

```html
<!doctype html>
<html>
<head>
...
		<link rel="stylesheet" href="../bower_components/medium-editor/dist/css/medium-editor.min.css" />
		<link rel="stylesheet" href="../bower_components/medium-editor/dist/css/themes/default.min.css" />
		<link href="../bower_components/font-awesome-5/css/all.min.css" rel="stylesheet" />
...
</head>
<body>
  <div class="editable"></div>

  <script type="text/javascript" src="<path_to_medium-editor>/dist/js/medium-editor.js"></script>
  <script type="text/javascript" src="<path_to_medium-editor-textcolor>/dist/js/medium-editor-textcolor.js"></script>

  <script type="text/javascript" charset="utf-8">
  var editor = new MediumEditor('.editable', {
    buttonLabels: 'fontawesome',
    toolbar: {
      buttons: [
        'bold',
        'italic',
        'textcolor'
      ]
    },
    extensions: {
      colorPicker: mediumtextcolor.createInstance()
    }
  });
  </script>
</body>
</html>
```

## Initialization options

* __showHSL__: 	If falsy hide the HSL inputs. Default: 	true.
* __showRGB__: 	If falsy hide the RGB inputs. Default: true.
* __showHEX__: 	If falsy hide the RGB HEX inputs. Default: true.
* __showAlpha__: 	If truly show control to change opacity. Default: false.
* __color__: 	Initial color. Default: #ff0000.
* __palette__: Array of predefined colors.
Each element can be color name, rgb array, hex rgb, rgb(), rgba(), hsl(), hsla(). Type: Array, Default: null.
* __paletteEditable__: 	If truly make the palette editable. Default: false.
* __useAlphaInPalette__: If truly enable alpha channel in palette colors,
if 'auto' depends on the value of 'showAlpha'. Type:String|Boolean, Default: auto.
* __hueBarSize__: 	Size of hue bar in pixel (width, height). Type:Array, Default: [150,11].
* __slBarSize__: Size of saturation/luminance canvas in pixel (width, height).  Type:Array, Default: [232,150].

> about property detail look at [https://narsenico.github.io/a-color-picker/#api_options]

* __contentDefault__: when buttonLabels is not fontawesome,the Medium Editor toolbar button content
* __contentFA__: 	when buttonLabels is fontawesome,the Medium Editor toolbar button content
* __aria__: The button title in Medium Editor toolbar
* __removeButtonContent__: 	remove button content in a-color-picker content
* __removeButtonTitle__: 	remove button title in a-color-picker


### Example

```javascript
...
    extensions: {
      colorPicker: mediumtextcolor.createInstance()
    }
...
```

## Demo

Clone the repository and:

```
bower install
open demo/index.html
```
## dependency

a-color-picker:https://github.com/narsenico/a-color-picker

## License

MIT: https://github.com/sonce/medium-editor-textcolor/blob/master/LICENSE


CHANGELOG.md
# 变更日志

## 0.1.3 / 2020-4-11

- 支持在VSC调试
- 解决unresolve dependencies

## 0.1.2 / 2020-3-19

- 增加jsdom，让测试支持document和windows的访问

## 0.1.1 / 2020-3-17

- 支持Typescript编写测试用例

## 0.1.0 / 2020-3-17

- 项目初始化
- 支持 JS ,TS开发项目，通过更改package.json里面的srctype属性切换语言
- 更新所有的package到最新版


src/index.ts
import Builder from './Builder';
import { option } from './option';

export default class colorPicker extends MediumEditor.Extension {
	private constructor(option: option) {
		super();
		this.option = { ...this.defaultOption, ...option };
		this.aria = this.option.aria;
		this.contentDefault = this.option.contentDefault;
		this.contentFA = this.option.contentFA;
	}

	private option: option;
	private defaultOption: option = {
		showHSL: false,
		showRGB: false,
		showHEX: true,
		showAlpha: false,
		color: '#ff0000',
		paletteEditable: false,
		useAlphaInPalette: 'auto',
		hueBarSize: [150, 11],
		slBarSize: [232, 150],
		contentDefault: '<b>TC</b>',
		aria: 'Set Text Color',
		contentFA: '<i class="fas fa-fill"></i>',
		removeButtonContent: '<i class="fas fa-eraser"></i>',
		removeButtonTitle: 'Remove Color Style'
	};

	name = 'textcolor';
	aria = 'Set Text Color';
	action = 'textcolor';
	contentDefault = '<b>TC</b>';
	contentFA = '<i class="fas fa-fill"></i>';

	private builder: Builder;

	static createInstance(option: option = {}): unknown {
		const colorPickerExtension = MediumEditor.extensions.form.extend(new colorPicker(option));
		const colorPickerEx = new colorPickerExtension();
		return colorPickerEx;
	}

	handleClick = function (this: colorPicker, event: MouseEvent): void {
		event.preventDefault();
		event.stopPropagation();

		if (this.isActive()) {
			this.hide();
			return;
		}
		this.selectionState = this.base.exportSelection();

		// If no text selected, stop here.
		if (this.selectionState && this.selectionState.end - this.selectionState.start === 0) {
			return;
		}
		this.show();
	};

	hide = function (this: colorPicker): void {
		this.setInactive();
		this.builder.hide();
	};

	show = function (this: colorPicker): void {
		this.setActive();

		const range = MediumEditor.selection.getSelectionRange(this.document);
		const parentElement = MediumEditor.selection.getSelectedParentElement(range) as HTMLElement;

		this.builder.show(parentElement.style.color, this.button.offsetLeft);
	};

	getForm = function (this: colorPicker): HTMLElement {
		if (!this.builder) {
			this.builder = new Builder(this.document, this.option, (color?: string) => {
				this.base.importSelection(this.selectionState, false);
				this.document.execCommand('styleWithCSS', false, 'true');
				if (typeof color === 'undefined' || color == '') this.document.execCommand('removeFormat', false, 'foreColor');
				else this.document.execCommand('foreColor', false, color);
			});

			this.base.subscribe('hideToolbar', () => {
				this.hide();
			});
		}

		return this.builder.getElement();
	};

	// getTemplate = function (this: colorPicker): void {
	// 	debugger;
	// 	const template = [
	// 		'<input type="text" class="medium-editor-toolbar-input" placeholder="',
	// 		this.placeholderText,
	// 		'">'
	// 	];

	// 	template.push(
	// 		'<a href="#" class="medium-editor-toolbar-save">',
	// 		this.getEditorOption('buttonLabels') === 'fontawesome' ? '<i class="fas fa-fill"></i>' : 'C',
	// 		'</a>'
	// 	);
	// };
}
